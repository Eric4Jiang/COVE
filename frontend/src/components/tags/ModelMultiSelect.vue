<template>
	<div class="model-multi-select">
		
		<input type="text" 
			v-model="cleanedQuery"
			v-on:keydown.enter.prevent="selectTagOrNew()"
			v-on:focus="showDropdown()"
			v-on-clickaway="hideDropdown"
		>
		
		<ul id="filtered-list" v-bind:class="{ hidden: isHidden }">
			<li v-for="(model, index) in filteredTags" :class="{ 'active': index == 0 }">
				<div v-on:click.self="selectModel(model, $event)">{{model.name}}</div>
			</li>
		</ul>

		<ul id="selected-list">
			<li v-for="model in selectedTags">
				<div class="selected" v-on:click.self="unselectModel(model, $event)">{{model.name}}</div>
			</li>
		</ul>
	</div>
</template>

<script>
import { mixin as clickaway } from 'vue-clickaway';

export default {
	name: 'ModelMultiSelect',
	props: {
		models: Array,
		currentTags: {
      type: Array,
      default: () => {return []}
    },
		createNew: Boolean,
		category: String,
	},
	mixins: [ clickaway ],

	data () {
		return {
			query: '',
			removedTags: [],
			newlySelectedTags: [],
			isHidden: true,
		}
	},

	computed: {

		cleanedQuery: {
			get: function () { return this.query },
			set: function (newQuery) { this.query = newQuery.toLowerCase().replace(/[^a-z\s]/g,'') }
		},

		selectedTags () {
			// current tags plus any new tags minus removed tags
			return this.currentTags.concat(this.newlySelectedTags).filter((item) => !this.removedTags.includes(item))
		},

		notSelectedTags () {
			return this.models.filter((item) => !this.selectedTags.includes(item))
		},

		filteredTags () {
			if (this.query == '') { return this.notSelectedTags }
			return this.notSelectedTags.filter((item) => item.name.includes(this.query))
		}
	},

	methods: {

		resetComponent () {
			this.query = ''
			this.isHidden = true
		},
		
		// All changes are routed through selectModel and unselectModel
		selectModel(model) {
			this.newlySelectedTags.push(model)
			this.$emit('changedTags', this.selectedTags, this.category)
		},

		unselectModel(model) {
			this.removedTags.push(model)
			this.$emit('changedTags', this.selectedTags, this.category)
		},

		selectTagOrNew () {
			if (this.filteredTags.length == 0 && this.createNew) {
				// Create new
				this.selectModel({
					"name": this.query,
					"category": this.category,
					"new":true
				})
			} else if (this.filteredTags[0] != undefined) {
				// Select First value in unselected list
				this.selectModel(this.filteredTags[0])
			}

			this.resetComponent()
		},
		
		showDropdown () { this.isHidden = false },
		hideDropdown: function () { this.isHidden = true } // Must use 'function' for click-away mixin	
	},
}

</script>

<style scoped lang="scss">

$primary-width: 200px;

.hidden {
	display: none;
}

input {
	width: $primary-width;
	margin-bottom: 0;
}

ul {
	padding: 0;
	margin: 0;
	list-style: none;

	li {
		font-family: 'Open Sans', san-serif;
		font-size: 12px;
	}
}

.model-multi-select {

	#selected-list {
		display: flex;
		flex-wrap: wrap;

		.selected {
			display: inline-block;
			border: 1px solid #000;
			padding: 5px 10px;
			margin: 10px 10px 0 0;
		}
	}

	.selected:hover {
		background: #de6868;
	}

	#filtered-list {
		position: absolute;
		z-index: 100;
		background: #fff;

		li {
			border: 1px solid #000;
			border-bottom: none;
			width: $primary-width;
			padding: 5px 10px;
			cursor: pointer;
		}

		li:hover, .active {
			background: #eee;
		}

		li:first-child {
			border-top: none;
		}

		li:last-child {
			border-bottom: 1px solid #000;
		}
	}

	#filtered-list:hover {
		.active {
			background: auto;
		}
	}
}
	
</style>