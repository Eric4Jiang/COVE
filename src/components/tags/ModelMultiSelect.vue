<template>
	<div class="model-multi-select">
		
		<input type="text" 
			v-model="modelSearchString" 
			v-on:input="searchModelsList()"
			v-on:keydown.enter.prevent="selectModelOrNew()"
			v-on:focus="makeVisible()"
			v-on-clickaway="makeHidden"
		>
		
		<ul id="not-selected-list" v-bind:class="{ hidden: isHidden }">
			<li v-for="(model, index) in notSelectedModels" :class="{ 'active': index == 0 }">
				<div v-on:click.self="selectModel(model, $event)">{{model.name}}</div>
			</li>
		</ul>

		<ul id="selected-list">
			<li v-for="model in selectedModels">
				<div class="selected" v-on:click.self="unSelectModel(model, $event)">{{model.name}}</div>
			</li>
		</ul>
	</div>
</template>

<script>
import removeFromList from "@/components/utils.js"
import { mixin as clickaway } from 'vue-clickaway';

export default {
	name: 'ModelMultiSelect',
	props: {
		models: Array,
		createNew: Boolean,
		category: String
	},
  mixins: [ clickaway ],

	data () {
		return {
			selectedModels: [],
			filteredModels: [],
			modelSearchString: '',
			newModels: [], // Expects list of objs with name property
			isHidden: true,
		}
	},

	computed: {
		notSelectedModels () {
			// Everything that is not selected or filtered out by the search
			let notSelected = this.models.filter((item) => !this.selectedModels.includes(item))
			return notSelected.filter((item) => !this.filteredModels.includes(item))
		},

		toFilterModels () {
			return this.models.filter((item) => !this.selectedModels.includes(item))
		}
	},

	methods: {

		selectModel(model) {
			// notSelectedModels is computed - will adjust
			this.selectedModels.push(model)
			this.$store.commit('addSelectedTag', model)
		},

		unSelectModel(model) {
			// notSelectedModels is computed - will adjust
			this.selectedModels = removeFromList(model, this.selectedModels)
			this.$store.commit('removeSelectedTag', model)
		},
		
		searchModelsList() {
			this.modelSearchString = this.modelSearchString.toLowerCase()

			if (this.modelSearchString == '') { 
				this.filteredModels = []
			} else {
				this.filteredModels = this.toFilterModels.filter((item) => !item.name.includes(this.modelSearchString))
			}
		},

		selectModelOrNew () {
			this.modelSearchString = this.modelSearchString.toLowerCase()

			if (this.notSelectedModels.length == 0 && this.createNew) {
				// Create new
				this.selectModel({
					"name": this.modelSearchString,
					"category": this.category,
					"new":true
				})
				// Reset
				this.modelSearchString = ''
				this.filteredModels = []
			} else {
				// Select First value in unselected list
				if (this.notSelectedModels[0] != undefined) {
					this.selectModel(this.notSelectedModels[0])
					// Clear text
					this.modelSearchString = ''
				}
			}
		},
		
		makeVisible () {
			this.isHidden = false
		},

		makeHidden: function () {
			this.isHidden = true
		}		

	},

	beforeMount () {
		this.selectedModels = this.$store.state.selectedTags.filter((item) => {return item.category == this.category})
	}
}

</script>

<style scoped lang="scss">

$primary-width: 200px;

.hidden {
	display: none;
}

input {
	width: $primary-width;
	margin-bottom: 0;
}

ul {
	padding: 0;
	margin: 0;
	list-style: none;

	li {
		font-family: 'Open Sans', san-serif;
		font-size: 12px;
	}
}

.model-multi-select {

	#selected-list {
		display: flex;
		flex-wrap: wrap;

		.selected {
			display: inline-block;
			border: 1px solid #000;
			padding: 5px 10px;
			margin: 10px 10px 0 0;
		}
	}

	.selected:hover {
		background: #de6868;
	}

	#not-selected-list {
    position: absolute;
    background: #fff;

		li {
			border: 1px solid #000;
			border-bottom: none;
			width: $primary-width;
			padding: 5px 10px;
			cursor: pointer;
		}

		li:hover, .active {
			background: #eee;
		}

		li:first-child {
			border-top: none;
		}

		li:last-child {
			border-bottom: 1px solid #000;
		}
	}

	#not-selected-list:hover {
		.active {
			background: auto;
		}
	}
}
	
</style>