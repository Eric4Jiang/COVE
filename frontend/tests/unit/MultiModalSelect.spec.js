import { shallowMount } from '@vue/test-utils' // be sure to skip a line to avoid unit testing error

import ModelMultiSelect from '@/components/tags/ModelMultiSelect.vue'

let wrapper = null

beforeEach(() => {
    wrapper = shallowMount(ModelMultiSelect, {
        propsData: {
            models: [
                {
                    category: "tasks",
                    datasets: [],
                    name: "testing",
                    id: "5"
                }
            ]
        }
    })
})

afterEach(() => {
    wrapper.destroy()
})

describe('ModelMultiSelect', () => {
    describe(':initial state', () => {
        it('isHidden: filtered-list should be hidden', () => {
            let filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)
            expect(filtered_list.classes()).toContain('hidden')
        })

        it('filtered-list and selected-list initialized correctly', () => {
            wrapper.setProps({
                models: [
                    {
                        category: "tasks",
                        name: "test 1",
                        id: "5"
                    },
                    {
                        category: "tasks",
                        name: "test 2",
                        id: "6"
                    }
                ],
                currentTags: [
                    {
                        category: "tasks",
                        name: "test 3",
                        id: "1"
                    }
                ]
            })
            let filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)
            expect(filtered_list.findAll('li').length).toEqual(2)
            let selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
            expect(selected_list.findAll('li').length).toEqual(1)
        })
    })

    describe('@events', () => {
        it('focus on text box: removes `hidden` class on filtered-list', () => {
            const text = wrapper.findAll('[data-test-id="text"]').at(0)
            const filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)

            expect(filtered_list.classes()).toContain('hidden')
            text.trigger('focus')
            expect(filtered_list.classes()).not.toContain('hidden')
        })

        // NOT FINISHED. find a way to trigger vue-on-clickaway
        it('clickaway from box: adds `hidden` class on filtered-list', () => {
            const text = wrapper.findAll('[data-test-id="text"]').at(0)
            const filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)

            text.trigger('focus')
            expect(filtered_list.classes()).not.toContain('hidden')
            // text.trigger('clickaway')
            // expect(filtered_list.classes()).toContain('hidden')
            // expect(wrapper.element).toMatchSnapshot()
        })

        it('clicking tag on filteredTags adds tag to selectedTags', () => {
            // First check that filtered-list is populated and selected-list is not
            let filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)

            // Select (click) the tag from the filtered list
            let filtered_tag = filtered_list.findAll('li').at(0).find('div')
            expect(wrapper.element).toMatchSnapshot()
            filtered_tag.trigger('click')
            expect(wrapper.element).toMatchSnapshot()

            // Check that filtered-list is now empty and selected-list is populated
            let selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
            expect(selected_list.findAll('li').length).toEqual(1)

            filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)
            expect(filtered_list.findAll('li').length).toEqual(0)
        })

        it('clicking tag in selectedTags removes tag from selectedTags', () => {
            wrapper.setProps({
                models: [],
                currentTags: [
                    {
                        category: "tasks",
                        datasets: [],
                        name: "Airplane",
                        id: "7"
                    }
                ]
            })
            // filtered-list should be empty, and selectedTags should have one
            let filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)
            expect(filtered_list.findAll('li').length).toEqual(0)

            let selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
            expect(selected_list.findAll('li').length).toEqual(1)

            // click selected tag
            let selected_tag = selected_list.findAll('li').at(0).find('div')
            expect(wrapper.element).toMatchSnapshot()
            selected_tag.trigger('click')
            expect(wrapper.element).toMatchSnapshot()

            // Check that selected-list is now empty. The tag should be in removed_tags
            selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
            expect(selected_list.findAll('li').length).toEqual(0)
        })

        describe('@enter a word in text box', () => {
            it('create new tag', () => {
                wrapper.setProps({ createNew: true })
                wrapper.setData({query: 'birds'})
                let selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
                expect(selected_list.findAll('li').length).toEqual(0)

                expect(wrapper.element).toMatchSnapshot()
                const text = wrapper.findAll('[data-test-id="text"]').at(0)
                text.trigger('keydown.enter')
                expect(wrapper.element).toMatchSnapshot()

                selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
                expect(selected_list.findAll('li').length).toEqual(1)
            })

            it('select existing tag', () => {
                wrapper.setData({query: 'testing'})

                const text = wrapper.findAll('[data-test-id="text"]').at(0)
                text.trigger('keydown.enter')

                let filtered_list = wrapper.findAll('[data-test-id="filtered-list"]').at(0)
                expect(filtered_list.findAll('li').length).toEqual(0)
                let selected_list = wrapper.findAll('[data-test-id="selected-list"]').at(0)
                expect(selected_list.findAll('li').length).toEqual(1)
            })
        })
    })
})