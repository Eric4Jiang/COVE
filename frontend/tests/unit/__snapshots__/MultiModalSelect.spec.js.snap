// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelMultiSelect @events @enter a word in text box create new tag 1`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  />
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  />
</div>
`;

exports[`ModelMultiSelect @events @enter a word in text box create new tag 2`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  >
    <li>
      <div>
        testing
      </div>
    </li>
  </ul>
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  >
    <li>
      <div
        class="selected"
      >
        birds
      </div>
    </li>
  </ul>
</div>
`;

exports[`ModelMultiSelect @events clicking tag in selectedTags removes tag from selectedTags 1`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  />
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  >
    <li>
      <div
        class="selected"
      >
        Airplane
      </div>
    </li>
  </ul>
</div>
`;

exports[`ModelMultiSelect @events clicking tag in selectedTags removes tag from selectedTags 2`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  />
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  />
</div>
`;

exports[`ModelMultiSelect @events clicking tag on filteredTags adds tag to selectedTags 1`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  >
    <li>
      <div>
        testing
      </div>
    </li>
  </ul>
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  />
</div>
`;

exports[`ModelMultiSelect @events clicking tag on filteredTags adds tag to selectedTags 2`] = `
<div
  class="model-multi-select"
>
  <input
    data-test-id="text"
    type="text"
  />
   
  <ul
    class="hidden"
    data-test-id="filtered-list"
    id="filtered-list"
  />
   
  <ul
    data-test-id="selected-list"
    id="selected-list"
  >
    <li>
      <div
        class="selected"
      >
        testing
      </div>
    </li>
  </ul>
</div>
`;
